src/resolution/snapshot.rs

```diff
@@ -803,11 +808,154 @@ fn name_without_path(name: &str) -> &str {
+    let package_id = NpmPackageId::from_serialized(value)?;
+    root_packages.insert(package_req, package_id.clone());
+  }
+
+  // now fill the packages except for the dist information
+  let mut packages = Vec::with_capacity(lockfile.content.npm.packages.len());
+  for (key, package) in &lockfile.content.npm.packages {
+    let id = NpmPackageId::from_serialized(key)?;
+
+    // collect the dependencies
+    let mut dependencies = HashMap::with_capacity(package.dependencies.len());
+    for (name, specifier) in &package.dependencies {
+      let dep_id = NpmPackageId::from_serialized(specifier)?;
+      dependencies.insert(name.clone(), dep_id);
+    }
+
+    packages.push(IncompletePackageInfo { id, dependencies });
+  }
+
+  Ok(IncompleteSnapshot {
+    root_packages,
+    packages,
+  })
+}
+
+/// Constructs [`ValidSerializedNpmResolutionSnapshot`] from the given [`Lockfile`].
+///
+/// You should call [`incomplete_snapshot_from_lockfile`] first to get an
+/// [`IncompleteSnapshot`] instance that's passed as the first argument for this
+/// function.
+pub async fn snapshot_from_lockfile(
+  incomplete_snapshot: IncompleteSnapshot,
+  api: &dyn NpmRegistryApi,
+) -> Result<ValidSerializedNpmResolutionSnapshot, SnapshotFromLockfileError> {
+  // fetch the package version information
+  let pkg_nvs = incomplete_snapshot
+    .packages
+    .iter()
+    .map(|p| p.id.nv.clone())
+    .collect::<Vec<_>>();
+  let get_version_infos = || {
+    FuturesOrdered::from_iter(pkg_nvs.iter().map(|nv| async move {
+      let package_info = api
+        .package_info(&nv.name)
+        .await
+        .map_err(SnapshotFromLockfileError::PackageInfoLoad)?;
+      package_info
+        .version_info(nv)
+        .map_err(SnapshotFromLockfileError::VersionNotFound)
+    }))
+  };
+  let mut version_infos = get_version_infos();
+  let mut i = 0;
+  let mut packages = Vec::with_capacity(incomplete_snapshot.packages.len());
+  while let Some(result) = version_infos.next().await {
+    match result {
+      Ok(version_info) => {
+        packages.push(SerializedNpmResolutionSnapshotPackage {
+          id: incomplete_snapshot.packages[i].id.clone(),
+          dependencies: incomplete_snapshot.packages[i].dependencies.clone(),
+          dist: version_info.dist,
+          system: NpmResolutionPackageSystemInfo {
+            cpu: version_info.cpu,
+            os: version_info.os,
+          },
+          optional_dependencies: version_info
+            .optional_dependencies
+            .into_keys()
+            .collect(),
+        });
+      }
+      Err(err) => {
+        if api.mark_force_reload() {
+          // reset and try again
+          version_infos = get_version_infos();
+          i = 0;
+          packages.clear();
+          continue;
+        } else {
+          return Err(err);
+        }
+      }
+    }
+
+    i += 1;
+  }
+
+  // clear the cache to reduce space for cache
+  api.clear_cache();
```

Comment:
I think we can remove this from here and instead call it after running this function in the CLI

Reply:
Sounds good, now `clear_cache` method is removed from the trait, and it's moved to the CLI code: https://github.com/denoland/deno/pull/20024/files#diff-747e5bf0658bef1ca03cfa6b83c21b7ffc899ce4324557b2cc94356681e76699R841